File: perf-results/cross-lang/benchmarks/spectralnorm.rb
def a(i, j)
  1.0 / ((i + j) * (i + j + 1) / 2 + i + 1)
end

def av(x, y, n)
  n.times do |i|
    val = 0.0
    n.times { |j| val += a(i, j) * x[j] }
    y[i] = val
  end
end

def atv(x, y, n)
  n.times do |i|
    val = 0.0
    n.times { |j| val += a(j, i) * x[j] }
    y[i] = val
  end
end

def atav(x, y, n)
  u = Array.new(n, 0.0)
  av(x, u, n)
  atv(u, y, n)
end

n = (ARGV[0] || 100).to_i
u = Array.new(n, 1.0)
v = Array.new(n, 0.0)

10.times do
  atav(u, v, n)
  atav(v, u, n)
end

vBv = vv = 0.0
n.times do |i|
  vBv += u[i] * v[i]
  vv += v[i] * v[i]
end

printf("%.9f\n", Math.sqrt(vBv / vv))

File: perf-results/cross-lang/benchmarks/spectralnorm.py
import sys
from math import sqrt

def A(i, j):
    return 1.0 / ((i + j) * (i + j + 1) // 2 + i + 1)

def Av(x, y, n):
    for i in range(n):
        y[i] = sum(A(i, j) * x[j] for j in range(n))

def Atv(x, y, n):
    for i in range(n):
        y[i] = sum(A(j, i) * x[j] for j in range(n))

def AtAv(x, y, n):
    u = [0.0] * n
    Av(x, u, n)
    Atv(u, y, n)

def main():
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 100
    u = [1.0] * n
    v = [0.0] * n
    for _ in range(10):
        AtAv(u, v, n)
        AtAv(v, u, n)
    vBv = vv = 0.0
    for ue, ve in zip(u, v):
        vBv += ue * ve
        vv += ve * ve
    print("%.9f" % sqrt(vBv / vv))

main()
